README.md
# paleofetch-bsd

A rewrite of [neofetch](https://github.com/dylanaraps/neofetch) started by [ss7m for linux](https://github.com/ss7m/paleofetch).

Not much code was left, but name is beautiful and similar logic.

## IMPORTANT

This is developed on M1 arm64. It should work on x86_64, if not make an issue. No support for powerpc, but maybe will add it.

## Why use paleofetch over neofetch?

One major reason is the performance improvement.

At least for now paleofetch execution time is around 0.05s while neofetch will execute at least a second.

The plan is to support all BSDs(FreeBSD, NetBSD, OpenBSD, maybe Dragonfly), but for now only macOS.

![example output](.gitlab/example.png)

## Dependencies

Having a Mac, or one of the BSDs installed, you could be very helpful.

## Recaching

Paleofetch uses caching to run faster. To recache data run `paleofetch -r`

## Compiling

Enter folder and type `make` (can be bmake, gnumake also)

There is also option `make clean` to clean and `make dump` to make assembly code for debuging

## FAQ

**Q**: Do you really run neofetch every time you open a terminal?  

**ss7m**: Yes, I like the way it looks and like that it causes my prompt to start midway
down the screen. I do acknowledge that the information it presents is not actually useful.

**DB**: Yea, me too that's why I forked it.


CHANGELOG.md
### v0.5c - 2022-06-07
* Fixed 5 memory leaks(mostly functions that returned pointers, not being assigned and freed)

### v0.5b - 2022-04-30
* Added usage prompt
* Removed some unnecessary stuff (typedef ulong etc)

### v0.5a - 2022-01-20
* Removed many unnecessary allocations, instead copying function results to one String located in main
* Some work is being done to support IOKit calls more easily, but M1 seems different than Intel mac version(some headers are missing).
* battery on by default
* trying to make portable makefile, but if project grows this will be problematic

### v0.5 - 2022-01-09
* Fixed many possible memory leaks(memory is now freed)
* Added battery procentage for mac, but still problematic, for unknown reason
* Added malloc\_s which tries three times allocating if ptr = NULL, if not possible exits. 

### v0.4 - 2021-12-05
* Added Caching
* Tried to simplify code when possible.

### v0.3b - 2021-11-27
* merge get\_sysctl\_info\_int and get\_sysctl\_info\_str into one.
* fixed date printing two spaces, when minutes are equal to zero.

### v0.3 - 2021-11-22
* reworked program to work more like original(author was way smarter than me)
* Added printing labels by default

### v0.2b - 2021-11-02
* support for m1, m1 gpu etc
* fixed memory on m1(hopefully works on intel64 as well)
* added days in uptime

### v0.2a - 2021-09-16
* mac gpu support
* Cleaned code, moved macOS only code to macintosh.c
* Added architecutre to OS name

### v0.1f - 2021-04-21
* Managed to get system resolution from Cocoa, paleofetch is much much faster now

### v0.1e - 2021-04-20
* Changed licence, basic functionality achieved, now development will freeze, and I will learn math.

### v0.1d - 2021-04-18
* Added speculative pages to used memory. Chnages in README

### v0.1c - 2021-04-17
* Fixed reading memory, finally logo is printed.

### v0.1b - 2021-04-14
* Implemented faster getting stats from vm.

### v0.1a - 2021-04-12
* Minimum Needed to show info. Still in monochorome and code is a mess. No Caching.


src/paleofetch.c
//
//  paleofetch.c
//  paleofetch-mac
//
//  Created by DesantBucie on 07/04/2021.
//
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <errno.h>
#include <sys/sysctl.h>

#include "config.h"
#include "paleofetch.h"
#include "sysctl_info.h"

#if defined(__MACH__) || defined(__APPLE__)
#include "macintosh.h"
#include "logos/macintosh_logo.h"
#endif

struct conf {
    char *label;    
    void (*function)();
    bool cached;
} config[] = CONFIG;

typedef struct {
    char str[256];
} String;

void get_colors1(char *colors1) {

    char *s = colors1;

    for(int i = 0; i < 8; i++) {
        sprintf(s, "\033[4%dm   ", i);
        s += 8;
    }
    strlcat(s, "\033[0m", 5);

}

void get_colors2(char *colors2) {

    char *s = colors2;

    for(int i = 8; i < 16; i++) {
        sprintf(s, "\033[48;5;%dm   ", i);
        s += 12 + (i >= 10 ? 1 : 0);
    }
    strlcat(s,"\033[0m", 5);
}
void get_uptime(char *ret_string)
{
    struct timeval boottime;
    size_t len = sizeof(struct timeval);
    int mib[2] = {CTL_KERN, KERN_BOOTTIME};
    int err = sysctl(mib, 2, &boottime, &len, NULL, 0);
    if (err != 0){
        halt_and_catch_fire("sysctl error", EXIT_FAILURE);
    }

    time_t const bsec = boottime.tv_sec, csec = time(NULL);
    float const time = difftime(csec, bsec) / 60;
    uint const days = time / (60 * 24);
    uint const hours = (uint)(time / 60) % 24;
    uint const minutes = (uint)time % 60;

    char days_string[BUFF_64];
    char hours_string[BUFF_64];
    char minutes_string[BUFF_64];

    if(days > 0)
        snprintf(days_string, BUFF_64, "%u %s", days, (days > 1 ? "days " : "day "));
    if(hours > 0)
        snprintf(hours_string, BUFF_64, "%u %s", hours, (hours == 0 || hours > 1 ? "hours " : "hour "));

    snprintf(minutes_string, BUFF_64, "%u %s", minutes, (minutes == 0 || minutes > 1 ? "minutes" : "minute"));
    snprintf(ret_string, BUFF_64 , "%s%s%s", days_string, hours_string, minutes_string);
    
}
void get_shell(char *shell)
{
        char *const shell_path = getenv("SHELL");
        // Manual says, don't touch original pointer
        char *s = shell_path;

        if(shell_path == NULL){
            strcpy(shell,"Unknown");
            return;
        }
        //After last '/' there will be name of the shell
        s = strrchr(s,'/'); 
        s++;
        strlcpy(shell,s, 8);
}
void hostname_underline(char *ret_string)
{
        // Composing <username>@<hostname> second time,
        // to properly calculate string lenght without ESC chars
    
        char userhost[BUFF_256];
        size_t const string_size = BUFF_256;

        snprintf(userhost, string_size, "%s%c%s", getenv("USER"), '@', details.nodename);
        size_t const underline = strlen(userhost);

        char *s = ret_string;
        for(size_t i = 0; i < underline; i++)
        {
                *s = '-';
                s++;
        }
        *s = '\0';
}
/* int check_for_pkg_info()
{
        int ret_int = 0;
        if(access("/opt/pkg/sbin/pkg_info"))
        {
                char *cmd = "pkg_info | wc -l";
                FILE *stdout_file = popen(cmd, "r");
                char *ret_str = malloc_s(BUFF_32);
                if (stdout_file)
                {
                        fgets(ret_str, BUFF_32, stdout_file);
                        pclose(stdout_file);
                }
                ret_int = atoi(ret_str);
        }
        else
        {
                printf("No pkg_info");
        }
        return ret_int;
}
 char *look_for_package_managers()
{
        char *packages = malloc_s(BUFF_256);
        int number = check_for_pkg_info();
        if(number != 0)
        {
                snprintf(packages, BUFF_256, "%d", check_for_pkg_info());
                strlcat(packages, "\40pkg_info\41", BUFF_256);
        }
        return packages;
}
*/
void get_user_and_host(char *userhost)
{
    snprintf(userhost, BUFF_256, "%s%s%s%s%s", "\033[1m", getenv("USER"), "\033[0m@\033[1m", details.nodename, "\033[0m");
}
void *cache_file_path(){
    char *const path = malloc_s(BUFF_256);
    if(getenv("XDG_CACHE_HOME")){
        snprintf(path, BUFF_256, "%s/.cache/paleofetch", getenv("HOME"));
    }
    else {
        snprintf(path, BUFF_256, "%s/.paleofetch", getenv("HOME"));
    }
    return path;
}
int check_cache_file(_Bool const recache)
{
    FILE *cache_file;
    String test;
    void *path = cache_file_path();
    if(fopen(path, "r") == NULL || recache == true)
    {   
        cache_file = fopen(path, "w"); 
        if(cache_file == NULL)
            return 1;
        for(uint i = 0; i < COUNT(config); i++){
            config[i].function(test.str);
            fprintf(cache_file,"%s|", test.str);
        }
        fclose(cache_file);
    }
    free(path);
    return 0;
}
char **get_cached_value(char **file_ret){
    void *path = cache_file_path();
    FILE *const cache_file = fopen(path, "r");
    free(path);
    if(cache_file == NULL){
        return NULL;
    }
    *file_ret = malloc_s(BUFF_512);
    fgets(*file_ret, BUFF_512, cache_file);
    fclose(cache_file);
    char **const list = malloc((COUNT(config)) * sizeof(char*));
    if(!list){
        free(file_ret);
        halt_and_catch_fire("Malloc error", 127);
    }
    char **list_ptr = list;
    char *file_ret_cpy = *file_ret; 
    while ((*list_ptr = strsep(&file_ret_cpy, "|")) != NULL) {
        list_ptr++; // strsep manual example 1:1
    }
    return list; 
}
void get_cpu(char *cpu){
    char *cpu_name = get_sysctlbyname_info_str(CPU);
    strlcpy(cpu, cpu_name, 256);
    free(cpu_name);
}
void get_terminal(char *terminal){
    if(getenv("TERM_PROGRAM") == NULL){
        strcpy(terminal, "Unknown");
        return; 
    }
    strlcpy(terminal, getenv("TERM_PROGRAM"), BUFF_256);
}
void spacer(char *spacer){
    strcpy(spacer, " ");
}
void get_machine(char *machine){
    char *model = get_sysctl_info(CTL_HW, HW_MODEL);
    strlcpy(machine, model, BUFF_256);
    free(model);
}
int main(int argc, char **argv)
{       
    
    int const ret = uname(&details);

    uint const logo_size = COUNT(logo);
    uint const config_size = COUNT(config);
    //sets which size we should base our iteration on, logo size or info size.
    uint const which_bigger = logo_size > config_size ? logo_size : config_size;

    if(argc == 1) {
        check_cache_file(false);
    }    
    else{ 
        if(!strcmp(argv[1], "-r"))
            check_cache_file(true);
        else {
            printf("Usage: paleofetch [-r (recache)]\n");
            return 0;
        }
    }
    char **file_ret = malloc(sizeof(char *));
    char **cached_list = get_cached_value(file_ret);
    String test;

    for(uint i = 0; i < which_bigger; i++)
    {
        // if i > count, we will print only logo, as we already are iterating on it's size.
        if(i >= config_size) {
            printf("%s\033[0m", logo[i]);
        }
        // If we run out of logo, but we have still info to print, we will have to 
        // print spaces corresponding to logos width, and then print our info.
        else if( i >= logo_size && i <= config_size) {
            for(int j = 0; j < logo_line_lenght; j++){
                printf("%c",' ');
            }
            printf("\033[1m%s\033[0m",config[i].label);
            if(config[i].cached == true && cached_list[i] != NULL){
                printf("%s", cached_list[i]);
            }
            else{
                config[i].function(test.str);
                printf("%s", test.str);
            }
        }
        // The only option left is that we have both info and logo to print.
        else { 
            printf("%s\033[0m ", logo[i]);
            printf("\033[1m%s\033[0m",config[i].label);
            if(config[i].cached == true && cached_list[i] != NULL){
                printf("%s", cached_list[i]);
            }
            else {
                config[i].function(test.str);
                printf("%s", test.str);
            }
        }

        puts("");
    }
    free(cached_list);
    free(*file_ret);
    free(file_ret);
    return ret;
}

